﻿﻿﻿﻿﻿﻿﻿﻿﻿﻿﻿﻿﻿﻿# **Data Inspection****Getting Started**﻿﻿﻿﻿﻿﻿﻿﻿﻿﻿﻿﻿Enter Directory `cd Desktop`Enter`cd BCB546X-Fall2018`Enter `cd assignments`Enter `cd UNIX_Assignment`﻿﻿﻿﻿**Inspection**﻿﻿﻿﻿﻿﻿﻿﻿﻿﻿﻿﻿﻿﻿﻿Command to look at data ` cut -f 1 -d ' ' chr1_increasing_SNPs_maize.txt | head -2` better command this cuts out the columns and shows the first 5 rows. ` grep -v "^#" order_joined_teosinte_genotypes.txt | cut -f 1-5 | head -n 5` or ` grep -v "^#" order_joined_teosinte_genotypes.txt | cut -f 1-5 | tail -n 5` to see the bottomFor checking columns and unique items cut -f 3 fang_et_al_genotypes.txt |sort| uniq -cTo check all of my files I used `ls` to view all of the necessary files for this assignment and confirm all files were present. Check File Size of all files in a directory `ls -l *`or I can check the size of a specific file `ls -l filename`To check the number of columns in a file use the command `awk '{print NF}' **filename** | sort -nu | tail -n 1` or  `head -n 1 filename | awk '{print NF}'`To check the number of lines us command `wc -l filename`﻿To look at the text of the files I am inspecting `cat fang_et_al_genotypes.txt` and `cat snp_position.txt`Lots of data and this is too much for us to comprehend# **Data Management****Maize**To sort the data and find the unique maize and teosinte lines in column 3 we use the command `cut -f3 fang_et_al_genotypes.txt | sort | uniq -c`maize: zmmil=290, zmmlr=1256, zmmmr=27 and teosinte: zmpba=900, zmpil=41, zmpja=34 All of the maize start with zmm… so I can `grep` which will grab all of the rows that have zmm from the genotype txt and move those rows to a new txt file `grep 'ZMM' fang_et_al_genotypes.txt > maize_zmmil_zmmlr_zmmmr.txt`Headers are not included so I will get those as well for the maize into a new file using the command  `head -n 1 fang_et_al_genotypes.txt > header_fang_et_al_genotypes.txt` to get the headers into a new file Now to combine or concatenate the header and the maize genotypes `cat header_fang_et_al_genotypes.txt maize_zmmil_zmmlr_zmmmr.txt > header_maize_genotypes.txt` Now it is time to turn the rows of the maize data into the columns so it can be joined with the snp data. I was given this command from Prof. Hufford `awk -f transpose.awk header_maize_genotypes.txt > transposed_maize_genotypes.txt`According to the lecture we need to sort the data but the shortcut `sort -c` can be used to see if the data is already sorted `sort -c transposed_maize_genotypes.txt` shows the data is not sorted but how we can sort the data `tail -n +2 transposed_maize_genotypes.txt | sort -k1,1 > sorted_transposed_maize_genotypes.txt`**SNP**The maize genotypes are now ready to be joined so I can begin working on the SNP data. The sort -c command is used again to show if data is sorted. `sort -c snp_position.txt`It is not sorted so I use the command `sort -k1,1 snp_position.txt > sorted_snp_position.txt`The header is not included so I need to use similar steps as above to create new head file**Join the SNP and the Fang Maize Genotype Data**Command `join -t $'\t' sorted_snp_position.txt sorted_transposed_maize_genotypes.txt > joined_maize_genotypes.txt` allows the two files -1 and -2 to be joined based on the common 1 column and put into a new fileTo look at this new file I can use `grep -v "^#" joined_maize_genotypes.txt | cut -f 1-20 | head -n 5` to see I need to remove some lines to get the file into the SNP ID, Chr# and position. From this data I see that I want to keep columns 1,3,4 and 16 to 1586. I get 1586 by using the command `head -n1 joined_maize_genotypes.txt | awk '{print NF}'` to find the total number of lines and the commandAs a check of my work I can use  the command `cat ordered_joined_maize_genotypes.txt | cut -f 2 | sort | uniq -c`    155 1     53 10    127 2    107 3     91 4    122 5     76 6     97 7     62 8     60 9      6 multiple     27 unknownand the command `wc -l filename` as a check of my work**Individual Maize chromosome with SNP position increasing and missing data with ?**To extract the data to meet the criteria use a command that looks in the second column for a specific number and then piping that to sort from smallest to largest in the 3rd column before moving the data to a new file.  chromosome 1 `awk '$2 == 1' joined_maize_genotypes.txt | sort -k3,3n > chr1_increasing_SNPs_maize.txt`chromosome 2  `awk '$2 == 2' joined_maize_genotypes.txt | sort -k3,3n > chr2_increasing_SNPs_maize.txt` chromosome 3  `awk '$2 == 3' joined_maize_genotypes.txt | sort -k3,3n > chr3_increasing_SNPs_maize.txt`chromosome 4  `awk '$2 == 4' joined_maize_genotypes.txt | sort -k3,3n > chr4_increasing_SNPs_maize.txt`chromosome 5  `awk '$2 == 5' joined_maize_genotypes.txt | sort -k3,3n > chr5_increasing_SNPs_maize.txt`chromosome 6  `awk '$2 == 6' joined_maize_genotypes.txt | sort -k3,3n > chr6_increasing_SNPs_maize.txt`chromosome 7  `awk '$2 == 7' joined_maize_genotypes.txt | sort -k3,3n > chr7_increasing_SNPs_maize.txt`chromosome 8  `awk '$2 == 2' joined_maize_genotypes.txt | sort -k3,3n > chr8_increasing_SNPs_maize.txt`chromosome 9  `awk '$2 == 9' joined_maize_genotypes.txt | sort -k3,3n > chr9_increasing_SNPs_maize.txt`chromosome 10  `awk '$2 == 10' joined_maize_genotypes.txt | sort -k3,3n > chr10_increasing_SNPs_maize.txt`**Add Header**Use the basic command `cat file1 file2 > chr#_new_file.txt` to combine the txt file containing the header and combine it with the data file and create a new file with all the info`cat headersers.txt chr1_increasing_SNPs_maize.txt > chr1_increasing_SNPs_maize.txt` To show my work make new files for the rest `cat headers.txt chr2_increasing_SNPs_maize.txt > chr2_increasing_SNPs_maize_headers.txt``cat headers.txt chr3_increasing_SNPs_maize.txt > chr3_increasing_SNPs_maize_headers.txt``cat headers.txt chr4_increasing_SNPs_maize.txt > chr4_increasing_SNPs_maize_headers.txt``cat headers.txt chr5_increasing_SNPs_maize.txt > chr5_increasing_SNPs_maize_headers.txt``cat headers.txt chr6_increasing_SNPs_maize.txt > chr6_increasing_SNPs_maize_headers.txt``cat headers.txt chr7_increasing_SNPs_maize.txt > chr7_increasing_SNPs_maize_headers.txt``cat headers.txt chr8_increasing_SNPs_maize.txt > chr8_increasing_SNPs_maize_headers.txt``cat headers.txt chr9_increasing_SNPs_maize.txt > chr9_increasing_SNPs_maize_headers.txt``cat headers.txt chr10_increasing_SNPs_maize.txt > chr10_increasing_SNPs_maize_headers.txt`**Remove Unknown and Multiple**Command to remove unknowns and multiple from chromosome data. `grep` searches for our `-v` is a search pattern and `E` tells it to go to the end of the line `grep -vE "unknown | multiple" filename searched > new file` example command for chromosome 10 `grep -vE "unknown | multiple" chr10_increasing_SNPs_maize.txt > chr10_increasing_SNPs_maize_headers_only.txt`**Replace Data**Next the assignment asks me to replace the '?" with a "-" and order the SNPs based on decreasing position value which can be done with the `sed` command which will find the "?" and replace with "-" with the command `sed -e 's/?/-/g' joined_maize_genotypes.txt > maize_genotypes_unknowns_dash.txt` followed by similar commands to what was used abovechromosome 1 `awk '$2 == 1' maize_genotypes_unknowns_dash.txt | sort -k3,3nr > chr1_decreasing_SNPs_maize_dash.txt`chromosome 2 `awk '$2 == 2' maize_genotypes_unknowns_dash.txt | sort -k3,3nr > chr2_decreasing_SNPs_maize_dash.txt`chromosome 3 `awk '$2 == 3' maize_genotypes_unknowns_dash.txt | sort -k3,3nr > chr3_decreasing_SNPs_maize_dash.txt`chromosome 4 `awk '$2 == 4' maize_genotypes_unknowns_dash.txt | sort -k3,3nr > chr4_decreasing_SNPs_maize_dash.txt`chromosome 5 `awk '$2 == 5' maize_genotypes_unknowns_dash.txt | sort -k3,3nr > chr5_decreasing_SNPs_maize_dash.txt`chromosome 6 `awk '$2 == 6' maize_genotypes_unknowns_dash.txt | sort -k3,3nr > chr6_decreasing_SNPs_maize_dash.txt`chromosome 7 `awk '$2 == 7' maize_genotypes_unknowns_dash.txt | sort -k3,3nr > chr7_decreasing_SNPs_maize_dash.txt`chromosome 8 `awk '$2 == 8' maize_genotypes_unknowns_dash.txt | sort -k3,3nr > chr8_decreasing_SNPs_maize_dash.txt`chromosome 9 `awk '$2 == 9' maize_genotypes_unknowns_dash.txt | sort -k3,3nr > chr9_decreasing_SNPs_maize_dash.txt`chromosome 10 `awk '$2 == 10' maize_genotypes_unknowns_dash.txt | sort -k3,3nr > chr10_decreasing_SNPs_maize_dash.txt`**Add Headers** Reasoning same as above`cat headers.txt chr1_decreasing_SNPs_maize_dash.txt > chr1_decreasing_SNPs_maize_dash_head.txt``cat headers.txt chr2_decreasing_SNPs_maize_dash.txt > chr2_decreasing_SNPs_maize_dash_head.txt``cat headers.txt chr3_decreasing_SNPs_maize_dash.txt > chr3_decreasing_SNPs_maize_dash_head.txt``cat headers.txt chr4_decreasing_SNPs_maize_dash.txt > chr4_decreasing_SNPs_maize_dash_head.txt``cat headers.txt chr5_decreasing_SNPs_maize_dash.txt > chr5_decreasing_SNPs_maize_dash_head.txt``cat headers.txt chr6_decreasing_SNPs_maize_dash.txt > chr6_decreasing_SNPs_maize_dash_head.txt``cat headers.txt chr7_decreasing_SNPs_maize_dash.txt > chr7_decreasing_SNPs_maize_dash_head.txt``cat headers.txt chr8_decreasing_SNPs_maize_dash.txt > chr8_decreasing_SNPs_maize_dash_head.txt``cat headers.txt chr9_decreasing_SNPs_maize_dash.txt > chr9_decreasing_SNPs_maize_dash_head.txt``cat headers.txt chr10_decreasing_SNPs_maize_dash.txt > chr10_decreasing_SNPs_maize_dash_head.txt`**Remove Unknown and Multiple**Command to remove unknowns and multiple from chromosome data. `grep` searches for our `-v` is a search pattern and `E` tells it to go to the end of the line `grep -vE "unknown | multiple" filename searched > new file` example command for chromosome 10 `grep -vE "unknown | multiple" chr10_decreasing_SNPs_maize.txt > chr10_decreasing_SNPs_maize_head_removed.txt`**Multiple and Unknown Maize**The `awk` command used to get the chromosomes can be used to get the multiple and unknown`awk '$2=="multiple"' joined_maize_genotypes.txt > maize multiple.txt`add the headers `cat headers.txt maize_multiple.txt > headers_maize_multiple.txt`The awk command used to get the chromosomes can be used to get the multiple and unknown`awk '$2=="unknown"' joined_maize_genotypes.txt > maize unknown.txt`add the headers `cat headers.txt maize_unknown.txt > headers_maize_unknown.txt`**Teosinte**As a check of my work I can use the command`cat ordered_joined_teosinte_genotypes.txt | cut -f 2 | sort | uniq -c`         155 1     53 10    127 2    107 3     91 4    122 5     76 6     97 7     62 8     60 9      6 multiple     27 unknownand the command `wc -l filename` as a check of my workThe teosinte all start with ZMP so command `grep 'ZMP' fang_et_al_genotypes.txt > teosinte_zmpba_zmpil_zmpja.txt` allows us to pull data we want from fangThis data is missing the header so we can use the txt file with the header info we got earlier to `cat` the data together `cat header_fang_et_al_genotypes.txt teosinte_zmpba_zmpil_zmpja.txt > header_teosinte_genotypes.txt`The data can now be transposed with the given command `awk -f transpose.awk header_teosinte_genotypes.txt > transposed_teosinte_genotypes.txt` the command `sort -c` shows that this is out of alignment which is expected but once again we can see how to sort the data and use the command `tail -n +2 transposed_teosinte_genotypes.txt | sort -k1,1 > sorted_transposed_teosinte_genotypes.txt` to create a new sorted and transposed teosinte file. **Join the SNP and the Fang Teosinte Genotype Data**Similar to above the command `join -t $'\t' -1 1 -2 1 sorted_snp_position.txt sorted_transposed_teosinte_genotypes.txt > joined_teosinte_genotypes.txt` To look at this new file  use `grep -v "^#" joined_teosinte_genotypes.txt | cut -f 1-20 | head -n 5` to see I need to remove some lines to get the file into the SNP ID, Chr# and position. From this data I see that I want to keep columns 1,3,4 and 16 to 988. I get 988 by using the command `head -n1 joined_teosinte_genotypes.txt | awk '{print NF}'` to find the total number of lines and the command `cut -f1,3,4,16-1586 joined_teosinte_genotypes.txt > ordered_joined_teosinte_genotypes.txt` to give a file with the desired order and data to start arranging. chromosome 1 `awk '$2 == 1' ordered_joined_teosinte_genotypes.txt | sort -k3,3n > chr1_increasing_SNPs_teosinte.txt`chromosome 2  `awk '$2 == 2' ordered_joined_teosinte_genotypes.txt | sort -k3,3n > chr2_increasing_SNPs_teosinte.txt` chromosome 3  `awk '$2 == 3' ordered_joined_teosinte_genotypes.txt | sort -k3,3n > chr3_increasing_SNPs_teosinte.txt`chromosome 4  `awk '$2 == 4' ordered_joined_teosinte_genotypes.txt | sort -k3,3n > chr4_increasing_SNPs_teosinte.txt`chromosome 5  `awk '$2 == 5' ordered_joined_teosinte_genotypes.txt | sort -k3,3n > chr5_increasing_SNPs_teosinte.txt`chromosome 6  `awk '$2 == 6' ordered_joined_teosinte_genotypes.txt | sort -k3,3n > chr6_increasing_SNPs_teosinte.txt`chromosome 7  `awk '$2 == 7' ordered_joined_teosinte_genotypes.txt | sort -k3,3n > chr7_increasing_SNPs_teosinte.txt`chromosome 8  `awk '$2 == 2' ordered_joined_teosinte_genotypes.txt | sort -k3,3n > chr8_increasing_SNPs_teosinte.txt`chromosome 9  `awk '$2 == 9' ordered_joined_teosinte_genotypes.txt | sort -k3,3n > chr9_increasing_SNPs_teosinte.txt`chromosome 10  `awk '$2 == 10' ordered_joined_teosinte_genotypes.txt | sort -k3,3n > chr10_increasing_SNPs_teosinte.txt`**Add Header**`cat headers.txt chr1_increasing_SNPs_teosinte.txt > chr1_increasing_SNPs_teosinte_head.txt``cat headers.txt chr2_increasing_SNPs_teosinte.txt > chr2_increasing_SNPs_teosinte_head.txt``cat headers.txt chr3_increasing_SNPs_teosinte.txt > chr3_increasing_SNPs_teosinte_head.txt``cat headers.txt chr4_increasing_SNPs_teosinte.txt > chr4_increasing_SNPs_teosinte_head.txt``cat headers.txt chr5_increasing_SNPs_teosinte.txt > chr5_increasing_SNPs_teosinte_head.txt``cat headers.txt chr6_increasing_SNPs_teosinte.txt > chr6_increasing_SNPs_teosinte_head.txt``cat headers.txt chr7_increasing_SNPs_teosinte.txt > chr7_increasing_SNPs_teosinte_head.txt``cat headers.txt chr8_increasing_SNPs_teosinte.txt > chr8_increasing_SNPs_teosinte_head.txt``cat headers.txt chr9_increasing_SNPs_teosinte.txt > chr9_increasing_SNPs_teosinte_head.txt``cat headers.txt chr10_increasing_SNPs_teosinte.txt > chr10_increasing_SNPs_teosinte_head.txt`**Remove Unknown and Multiple**Command to remove unknowns and multiple from chromosome data. `grep` searches for our `-v` is a search pattern and `E` tells it to go to the end of the line `grep -vE "unknown | multiple" filename searched > new file` example command for chromosome 10 `grep -vE "unknown | multiple" chr10_increasing_SNPs_teosinte_head.txt > chr10_increasing_SNPs_teosinte_head_only`**Replace ? with -**Next the assignment asks me to replace the '?" with a "-" and order the SNPs based on decreasing position value which can be done with the `sed` command which will find the "?" and replace with "-" with the command `sed -e 's/?/-/g' ordered_joined_teosinte_genotypes.txt > teosinte_genotypes_unknowns_dash.txt` followed by similar commands to what was used abovechromosome 1 `awk '$2 == 1' teosinte_genotypes_unknowns_dash.txt | sort -k3,3nr > chr1_decreasing_SNPs_teosinte_dash.txt`chromosome 2 `awk '$2 == 2' teosinte_genotypes_unknowns_dash.txt | sort -k3,3nr > chr2_decreasing_SNPs_teosinte_dash.txt`chromosome 3 `awk '$2 == 3' teosinte_genotypes_unknowns_dash.txt | sort -k3,3nr > chr3_decreasing_SNPs_teosinte_dash.txt`chromosome 4 `awk '$2 == 4' teosinte_genotypes_unknowns_dash.txt | sort -k3,3nr > chr4_decreasing_SNPs_teosinte_dash.txt`chromosome 5 `awk '$2 == 5' teosinte_genotypes_unknowns_dash.txt | sort -k3,3nr > chr5_decreasing_SNPs_teosinte_dash.txt`chromosome 6 `awk '$2 == 6' teosinte_genotypes_unknowns_dash.txt | sort -k3,3nr > chr6_decreasing_SNPs_teosinte_dash.txt`chromosome 7 `awk '$2 == 7' teosinte_genotypes_unknowns_dash.txt | sort -k3,3nr > chr7_decreasing_SNPs_teosinte_dash.txt`chromosome 8 `awk '$2 == 8' teosinte_genotypes_unknowns_dash.txt | sort -k3,3nr > chr8_decreasing_SNPs_teosinte_dash.txt`chromosome 9 `awk '$2 == 9' teosinte_genotypes_unknowns_dash.txt | sort -k3,3nr > chr9_decreasing_SNPs_teosinte_dash.txt`chromosome 10 `awk '$2 == 10' teosinte_genotypes_unknowns_dash.txt | sort -k3,3nr > chr10_decreasing_SNPs_teosinte_dash.txt`**Add Header**Use the `cat` command to combine the headers and chr#_increasing_SNPs_maize and create a new file. `cat headers.txt chr1_decreasing_SNPs_teosinte_dash.txt > chr1_decreasing_SNPs_teosinte_head.txt``cat headers.txt chr2_decreasing_SNPs_teosinte_dash.txt > chr2_decreasing_SNPs_teosinte_head.txt``cat headers.txt chr3_decreasing_SNPs_teosinte_dash.txt > chr3_decreasing_SNPs_teosinte_head.txt``cat headers.txt chr4_decreasing_SNPs_teosinte_dash.txt > chr4_decreasing_SNPs_teosinte_head.txt``cat headers.txt chr5_decreasing_SNPs_teosinte_dash.txt > chr5_decreasing_SNPs_teosinte_head.txt``cat headers.txt chr6_decreasing_SNPs_teosinte_dash.txt > chr6_decreasing_SNPs_teosinte_head.txt``cat headers.txt chr7_decreasing_SNPs_teosinte_dash.txt > chr7_decreasing_SNPs_teosinte_head.txt``cat headers.txt chr8_decreasing_SNPs_teosinte_dash.txt > chr8_decreasing_SNPs_teosinte_head.txt``cat headers.txt chr9_decreasing_SNPs_teosinte_dash.txt > chr9_decreasing_SNPs_teosinte_head.txt``cat headers.txt chr10_decreasing_SNPs_teosinte_dash.txt > chr10_decreasing_SNPs_teosinte_head.txt`**Remove Unknown and Multiple**Command to remove unknowns and multiple from chromosome data. `grep` searches for our `-v` is a search pattern and `E` tells it to go to the end of the line `grep -vE "unknown | multiple" filename searched > new file` example command for chromosome 10 `grep -vE "unknown | multiple" chr10_decreasing_SNPs_teosinte_head.txt > chr10_decreasing_SNPs_teosinte_head_only.txt`**Multiple and Unknown Maize**The `awk` command used to get the chromosomes can be used to get the multiple and unknown`awk '$2=="multiple"' ordered_joined_teosinte_genotypes.txt > teosinte_multiple.txt`add the headers `cat headers.txt teosinte_multiple.txt > headers_teosinte_multiple.txt`The awk command used to get the chromosomes can be used to get the multiple and unknown`awk '$2=="unknown"' ordered_joined_teosinte_genotypes.txt > teosinte_unknown.txt`add the headers `cat headers.txt teosinte_unknown.txt > headers_teosinte_unknown.txt`**Misc crap**This was some of the code created by user served no purpose but user was afraid to delete in case it became uselful. `cat headers.txt chr1_increasing_SNPs_maize.txt > chr1_increasing_SNPs_maize_headers.txt``cat headers.txt chr1_increasing_SNPs_maize.txt > chr1_increasing_SNPs_maize_headers.txt``cat headers.txt chr2_increasing_SNPs_maize.txt > chr2_increasing_SNPs_maize_headers.txt``cat headers.txt chr3_increasing_SNPs_maize.txt > chr3_increasing_SNPs_maize_headers.txt``cat headers.txt chr4_increasing_SNPs_maize.txt > chr4_increasing_SNPs_maize_headers.txt``cat headers.txt chr5_increasing_SNPs_maize.txt > chr5_increasing_SNPs_maize_headers.txt``cat headers.txt chr6_increasing_SNPs_maize.txt > chr6_increasing_SNPs_maize_headers.txt``cat headers.txt chr7_increasing_SNPs_maize.txt > chr7_increasing_SNPs_maize_headers.txt``cat headers.txt chr8_increasing_SNPs_maize.txt > chr8_increasing_SNPs_maize_headers.txt``cat headers.txt chr9_increasing_SNPs_maize.txt > chr9_increasing_SNPs_maize_headers.txt``cat headers.txt chr10_increasing_SNPs_maize.txt > chr10_increasing_SNPs_maize_headers.txt`{ echo "id1,id2,id3,id4"; cat file.csv; } > newfile.csvsed -i 1i"id1,id2,id3,id4" file.csv > chr_headerawk '$2 == unknown' joined_maize_genotypes.txt > test.txtawk '$2=="unknown"' joined_maize_genotypes.txt > test.txt used to find unknowns   